Add callback-based IO event notifications

fede1024 opened the pull request:
See issue #1859.

I've added a test based on 0040. I'm not super happy with the sleep-based waiting loop, but even with a better approach I don't think we'd be able to reduce the execution time of the test by more than 2 seconds. Let me know if you have any suggestion.

fede1024 commented on src/rdkafka.h:
```diff
@@ -2370,13 +2370,34 @@ size_t rd_kafka_queue_length (rd_kafka_queue_t *rkqu);
  *
  * librdkafka will maintain a copy of the \p payload.
  *
+ * @remark IO and callback event triggering are mutually exclusive.
```
I think this should actually be "@remark FD and callback event..." since they are both IO events.

fede1024 replied:
Dropped.

edenhill commented on src/rdkafka.h:
```diff
@@ -2370,13 +2370,34 @@ size_t rd_kafka_queue_length (rd_kafka_queue_t *rkqu);
  *
  * librdkafka will maintain a copy of the \p payload.
  *
+ * @remark FD and callback IO event triggering are mutually exclusive.
```
IO in this context refers to the file descriptor triggering.
I don't think the callback based triggering should be named IO since it doesn't really do any IO or can't be directly integrated with an IO scheduler in the way the fd-based one can.

So please revert your IO -> FD changes, and go with the following nomenclature:
IO event = FD
callback event = callback

fede1024 replied:
I see, makes sense. Thank you.

edenhill commented on src/rdkafka.h:
```diff
@@ -2370,13 +2370,34 @@ size_t rd_kafka_queue_length (rd_kafka_queue_t *rkqu);
 				     const void *payload, size_t size);
 
+/**
+ * @brief Enable IO event triggering for queue using a callback.
```
drop "IO"

edenhill commented on src/rdkafka.h:
```diff
@@ -2370,13 +2370,34 @@ size_t rd_kafka_queue_length (rd_kafka_queue_t *rkqu);
+ *
+ * To remove event triggering call with \p event_cb = NULL.
+ *
+ * @remark FD and callback IO event triggering are mutually exclusive.
```
FD -> IO

edenhill commented on src/rdkafka.h:
```diff
@@ -2370,13 +2370,34 @@ size_t rd_kafka_queue_length (rd_kafka_queue_t *rkqu);
+ *         handle).
+ */
+RD_EXPORT
+void rd_kafka_queue_io_event_cb_enable (rd_kafka_queue_t *rkqu,
```
drop _io_

fede1024 replied:
I'm going to replace `io` with `cb`: `rd_kafka_queue_cb_event_enable` (similarly to `rd_kafka_queue_io_event_enable`). Please let me know if you would rather call it `rd_kafka_queue_event_cb_enable`.

edenhill commented on src/rdkafka_queue.c:
```diff
@@ -722,6 +724,46 @@ void rd_kafka_queue_io_event_enable (rd_kafka_queue_t *rkqu, int fd,
+/**
+ * @brief Enable or disable(event_cb==NULL) callback-based wake-ups for queue
+ */
+void rd_kafka_q_io_event_cb_enable (rd_kafka_q_t *rkq,
```
drop _io_

edenhill commented on src/rdkafka_queue.c:
```diff
@@ -722,6 +724,46 @@ void rd_kafka_queue_io_event_enable (rd_kafka_queue_t *rkqu, int fd,
+
+}
+
+void rd_kafka_queue_io_event_cb_enable (rd_kafka_queue_t *rkqu,
```
dito

edenhill commented on tests/0083-io_event_cb.c:
```diff
@@ -0,0 +1,210 @@
+/*
+ * librdkafka - Apache Kafka C library
+ *
+ * Copyright (c) 2012-2013, Magnus Edenhill
```
Set copyright to 2018

edenhill commented on tests/0083-io_event_cb.c:
```diff
@@ -0,0 +1,210 @@
+         */
+        while (recvd < msgcnt) {
+                TEST_SAY("Waiting for event\n");
+                int event_count = wait_event_cb(1 * wait_multiplier);
```
variable definitions must go at the top of the scope, before any function calls.

edenhill commented on tests/0083-io_event_cb.c:
```diff
@@ -0,0 +1,210 @@
+                TEST_ASSERT(event_count <= 1, "Event cb called %d times", event_count);
+
+                if (event_count == 1) {
+                        TEST_SAY("Events received: %d\n", event_count);
```
dito

edenhill commented on src/rdkafka.h:
```diff
@@ -2359,7 +2359,7 @@ size_t rd_kafka_queue_length (rd_kafka_queue_t *rkqu);
 
 /**
- * @brief Enable IO event triggering for queue.
+ * @brief Enable FD event triggering for queue.
```
This should still be IO

edenhill commented on src/rdkafka_queue.h:
```diff
@@ -281,6 +285,11 @@ void rd_kafka_q_io_event (rd_kafka_q_t *rkq) {
+        if (rkq->rkq_qio->event_cb) {
+                rkq->rkq_qio->event_cb(rkq->rkq_rk, rkq->rkq_qio->event_cb_opaque);
+                return;
+	}
```
tab -> 8 whitespaces

fede1024 replied:
Sorry I missed this.

edenhill commented on tests/CMakeLists.txt:
```diff
@@ -73,6 +73,7 @@ set(
     0078-c_from_cpp.cpp
     0079-fork.c
     0081-fetch_max_bytes.cpp
+    0083-io_event_cb.c
```
this file will need to be added to win32/tests/tests.vcxproj too

fede1024 replied:
Will do. I'll also rename the test to cb_event, according to the naming convention.

edenhill commented:
The branch is conflicting, rebase on latest master.

fede1024 commented:
I've rebased, squashed and pushed again. Let me know if you have any additional comment.

Is there any chance this could make it into 0.11.5 or is it too close to release?

edenhill commented:
Thank you!
